### PR TITLE
Refined LintFix API

### PR DESCRIPTION
**Pull request by jpy-git**
### Brief summary of the change made

This PR defines API methods for creating LintFix instances. Should provide a more natural way to interact with lint fixes and allows for docstrings in the methods to better highlight the action of the fix.

Things that have changed:
- Four new methods on LintFix class: create_before, create_after, replace, delete.
- Renamed "edit" fix type to "replace" because a) it better describes the action and b) there is an instance attribute called edit that would cause name collisions.
- Updated all rules to use these new methods (except L022 due to its method of determining the rule fix type to use).
- Added basic typing to parts of LintFix.
- made edit attribute only accept an iterable of BaseSegment rather than a single BaseSegment for greater clarity and to remove redundant code lines.

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/rules/base.py [hunk 1]
```diff
@@ -113,28 +113,32 @@ class LintFix:
 
     """
 
-    def __init__(self, edit_type, anchor: BaseSegment, edit=None):
+    def __init__(
+        self,
+        edit_type: str,
+        anchor: BaseSegment,
+        edit: Optional[Iterable[BaseSegment]] = None,
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/base.py:120]: This only accepts `Iterable[BaseSegment]` rather than `Union[BaseSegment, Iterable[BaseSegment]]` as before. Think this is cleaner and more well defined.

File: src/sqlfluff/core/rules/base.py [hunk 2]
```diff
@@ -113,28 +113,32 @@ class LintFix:
...
-        # Copy all the elements of edit to stop contamination.
-        # We're about to start stripping the position markers
-        # of some of the elements and we don't want to end up
-        # stripping the positions of the original elements of
-        # the parsed structure.
-        self.edit = copy.deepcopy(edit)
-        if self.edit:
+        if edit is not None:
+            # Coerce edit iterable to list
+            edit = list(edit)
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/base.py:135]: iterables all get coerced to lists to keep internal usage unchanged, but rule writers are free to provide whichever format of iterable they choose to the API 👍

File: src/sqlfluff/core/rules/base.py [hunk 3]
```diff
@@ -145,10 +149,10 @@ def __init__(self, edit_type, anchor: BaseSegment, edit=None):
                         "Developer Note: Edit segment found with preset position marker. "
                         "These should be unset and calculated later."
                     )
-                    seg.pos_marker = None
-        # Once stripped, we shouldn't replace any markers because
-        # later code may rely on them being accurate, which we
-        # can't guarantee with edits.
+                    seg.pos_marker = None  # type: ignore
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/base.py:152]: existing type error, previously this wasn't getting checked as this class wasn't type hinted.
Had to type ignore this line as this is being typed elsewhere as `PositionMarker`: https://github.com/sqlfluff/sqlfluff/blob/7fd452dbfb9ab8537b8ca6e92c9489ff04d6172b/src/sqlfluff/core/parser/segments/base.py#L123

I tried changing that to `Optional[PositionMarker]` but that raised even more type errors as other areas of the code base aren't taking into account that seg.pos_marker could be set to None here for BaseSegments.

Basically this PR changes nothing here, so we can leave as is and investigate afterwards 👍

File: src/sqlfluff/core/rules/base.py [hunk 4]
```diff
@@ -203,6 +207,30 @@ def __eq__(self, other):
...
+    ) -> "LintFix":
+        """Create edit segments before the supplied anchor segment."""
+        return cls("create_before", anchor, edit)
+
+    @classmethod
+    def create_after(
+        cls, anchor: BaseSegment, edit: Iterable[BaseSegment]
+    ) -> "LintFix":
+        """Create edit segments after the supplied anchor segment."""
+        return cls("create_after", anchor, edit)
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/base.py:232]: This is the main addition of the PR, quite happy with how simple it is and think it makes using this class in the rules a lot nicer.

N.B. "edit" is renamed to "replace" as that better describes the fix type and avoids naming collisions with the existing edit instance attribute.

**Review by jpy-git — COMMENTED** [review]
@barrywhart this is ready for review 😄 

Given that this also updates a lot of the rule files, I've called out the main changes for you in this review + the PR description 👍

**Review by barrywhart — APPROVED** [review]
Looks great overall! A few small suggestions (optional to address).

File: src/sqlfluff/core/rules/base.py [hunk 5]
```diff
@@ -203,6 +207,30 @@ def __eq__(self, other):
...
         return True  # pragma: no cover TODO?
 
+    @classmethod
+    def delete(cls, anchor: BaseSegment) -> "LintFix":
+        """Delete supplied anchor segment."""
+        return cls("delete", anchor)
+
+    @classmethod
+    def replace(cls, anchor: BaseSegment, edit: Iterable[BaseSegment]) -> "LintFix":
+        """Replace supplied anchor segment with the edit segments."""
```
[the remark this answers is not in the document]
Comment (jpy-git) [reply src/sqlfluff/core/rules/base.py:217]: @barrywhart actually that's not how replace works. Refer here: https://github.com/sqlfluff/sqlfluff/blob/bb7fbd4c2c0d485079853846b1d306549013058e/src/sqlfluff/core/parser/segments/base.py#L938

This is how the "replace" (previously "edit") type works (and has always worked):
Let's for example say at some level of the parse tree with have these sibling segments
`(segment1, segment2, segment3)`
we then have a `LintFix` to replace `segment1` with `segment_a` and `segment_b`.
`LintFix.replace(segment1, [segment_a, segment_b])`
The fix gets applied at the SAME level of the parse tree as the anchor segment!
so you get
`(segment1, segment_a, segment_b, segment3)`
(all lint fixes get applied to the same level as the anchor segment)

You can also see why I call it replace rather than edit.

[the remark this answers is not in the document]
Comment (jpy-git) [reply src/sqlfluff/core/rules/base.py:217]: The key point to take away regarding incorrect parse trees from fixes is:
It is the PLACEMENT (i.e. `anchor_segment`) of the fix that causes these issues NOT the type of the fix.

example:
A rule that adds a whitespace after an identifier:
consider a tree like this:
```
table_identifier
|__(object_identifier_a, dot, object_identifier_b)
```
some rule then determines there should be a whitespace after the table identifier but since its maybe using raw segments it uses object_identifier as it's fix anchor and does a create_after (because the issue is not anything specifically about edit segments): `LintFix.create_after(object_identifier_b, [WhitespaceSegment()])`
this would give a tree like:
```
table_identifier
|__(object_identifier_a, dot, object_identifier_b, whitespace_segment)
```
which is obviously wrong, and leads to later rules not working as expected.

The correct approach would be `LintFix.create_after(table_identifier, [WhitespaceSegment()])` to give:
table_identifier
|__(object_identifier_a, dot, object_identifier_b)
whitespace_segment
```
which would be the same parse tree as if you had re-parsed the fixed string

[the remark this answers is not in the document]
Comment (jpy-git) [reply src/sqlfluff/core/rules/base.py:217]: So the challenge is around how you determine the correct anchor_segment placement, not what edit type is used

File: src/sqlfluff/core/rules/base.py [hunk 6]
```diff
@@ -203,6 +207,30 @@ def __eq__(self, other):
             return False
         return True  # pragma: no cover TODO?
 
+    @classmethod
+    def delete(cls, anchor: BaseSegment) -> "LintFix":
+        """Delete supplied anchor segment."""
+        return cls("delete", anchor)
+
+    @classmethod
+    def replace(cls, anchor: BaseSegment, edit: Iterable[BaseSegment]) -> "LintFix":
```
[the remark this answers is not in the document]
Comment (jpy-git) [reply src/sqlfluff/core/rules/base.py]: updated these to anchor_segment and edit segments. It lines up with the docstrings and think it's pretty clear.

File: src/sqlfluff/core/rules/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpy-git) [reply src/sqlfluff/core/rules/base.py:219]: also here is not the place to further discuss the above, will create a separate issue for that, but suffice to say that won't make this suggested change 👍
